### ISSUE TITLE
Request for a method to get the current mouse position

### ISSUE THREAD

**Issue by Hazer** [games-on-whales/inputtino#6]
In Sunshine I have a method to get the current mouse position. Currently it's used mostly for macOS, but I'm using it for testing each platform's mouse integration too.

Could inputtino provide something like that so I can avoid having to deal directly with libinput?

Something like:

```cpp
util::point_t
  get_location(input_raw_t *raw) { // Sunshine method
    if (raw->mouse) {
      auto x = (*raw->mouse).get_absolute_x(); // Potential inputtino new API
      auto y = (*raw->mouse).get_absolute_y(); // Potential inputtino new API
      return { x, y };
    }
    return { 0, 0 };
  }
```

**Comment by ABeltramo** [comment]
I wouldn't do that for a couple reasons:
 - virtual devices don't need to track what the current status is, so I don't have that information around
 - It is so much better to test with `libinput`+`SDL`!
   - You can validate stuff externally, how would you know that the internal status is in fact what will be reported outside?
   - A single unit test should work across all platforms since they both support Win/Linux/OSX
   - You can easily copy all my tests, see: [tests/testLibinput.cpp](https://github.com/games-on-whales/inputtino/blob/stable/tests/testLibinput.cpp) and [tests/testJoypads.cpp](https://github.com/games-on-whales/inputtino/blob/stable/tests/testJoypads.cpp)


I'll keep this open if you want to plead your case :sweat_smile:

**Comment by Hazer** [comment]
@ABeltramo While I've been thinking in toying with SDL for the tests, I don't think libinput would be useful for Windows/macOS at all, I'm not sure I follow your reasoning here, isn't libinput for X/Wayland only?

Mostly I agree with you, but I won't be testing inputtino directly, I'll be testing the platf layer, I shouldn't know, nor care, what it's running below, be it windows, macos, legacy input, inputtino, whatever, just that I call platform method to move the mouse, and that it moved where I wanted. But yeah, I think I'll need to rethink the tests, the way I'm doing it's not great. At the moment, I can't manually control the instantiation of the platf namespace, it's kind of hard to test it, I can't provide mocks or custom versions of the dependencies. In macOS we actually need to know the current mouse position to correctly trace mouse input, but that's not relevant to inputtino, so I think I'll rethink this from the start.

The tests in case are, just so you know which tests I'm mentioning: https://github.com/LizardByte/Sunshine/blob/f396059b64691a01b6490c5119555b5601bc6bfd/tests/unit/test_mouse.cpp

And if you have any suggestion, I'll be happy to hear, I'm scratching my head on how to properly do it without proposing even bigger refactors to Sunshine atm...

**Comment by ABeltramo** [comment]
First off, yep sorry I'm not sure why, but I've confused `libinput` (linux only) with `libusb` (cross platform). Either way, I think you can get everything just by plugging `SDL` which is properly cross platform, widely used by a lot of apps/games and it should support all our devices.

Back to the topic, I think that a great cross platform high-level unit test for this would be to simulate a Moonlight control packet and then expect a device to reflect the state that is defined in that packet. I have a bunch of tests like this in Wolf, ex:

https://github.com/games-on-whales/wolf/blob/5fdea7273b8323c0373eedc8f7a244676705692a/tests/platforms/linux/input.cpp#L97-L124

Sunshine, unfortunately, has a lot of intertwined global state and it might not be as easy to setup for testing. If you want to properly test it you'll have to do some fairly large refactoring; there's no way around it.. 

For input, specifically, it might not be that bad though. I think you can target the methods under [platform/common.h](https://github.com/LizardByte/Sunshine/blob/5f6fe3319c1fb383069d6b5fee19aeb37a8d891e/src/platform/common.h#L687-L774) and basically instantiate the devices using `input_raw_t` and `client_input_raw_t`, then call some method against them (ex: `move_mouse()`) and then check that the device actually did that using `SDL`.

On a separate note, bear in mind that Github Actions Runners are limited; in Linux there's no support for `uinput`/`evdev` and the likes so you can't run those tests there. I run these tests locally and on a Github Runner that I selfhost where I've added those devices.

**Comment by Hazer** [comment]
> I have a bunch of tests like this in Wolf

Those look great!

> Sunshine, unfortunately, has a lot of intertwined global state and it might not be as easy to setup for testing. If you want to properly test it you'll have to do some fairly large refactoring; there's no way around it..

Yeah, that's kinda why I started breaking those huge namespaces into smaller namespaces, turning the namespace mess into smaller namespaces mess, to then refactor it later, but it's a long way from avoiding global states yet.

> I think you can target the methods under [platform/common.h](https://github.com/LizardByte/Sunshine/blob/5f6fe3319c1fb383069d6b5fee19aeb37a8d891e/src/platform/common.h#L687-L774)

That's where I'm going, those tests use it, but were I think I got wrong, was that I was trying to get the current cursor position from the platform implementation itself, which seems like a pretty bad idea in hindsight, but came from the idea that macOS currently uses it internally and I wanted to test this too, but nah, it was a wrong way to approach it. 

I think your explanation and code samples gave me a really good idea on how to do the next iteration, and learning to use `SDL` will be really useful to me.

> bear in mind that Github Actions Runners are limited; in Linux there's no support for uinput/evdev

That's a really important note, thanks! I'd fall for that trap.

I think we can close this, I've gathered enough information and I don't think I have any counter argument to keep this request.

Thank you very much!
